File: RootTools2/src/main/java/com/rarnu/tools/neo/comp/LoadingView.java
package com.rarnu.tools.neo.comp;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.util.AttributeSet;
import android.view.Gravity;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import com.rarnu.tools.neo.R;
import com.rarnu.tools.neo.utils.UIUtils;

public class LoadingView extends LinearLayout {

    private ProgressBar pb = null;
    private TextView tv = null;

    public LoadingView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initAttrs(attrs);
    }
    public LoadingView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initAttrs(attrs);
    }
    public LoadingView(Context context) {
        super(context);
        initAttrs(null);
    }

    private void initAttrs(AttributeSet attrs) {
        setOrientation(HORIZONTAL);
        setBackground(getContext().getResources().getDrawable(R.drawable.background_layout, getContext().getTheme()));

        // inner component
        pb = new ProgressBar(getContext());
        LinearLayout.LayoutParams rllpPb = new LinearLayout.LayoutParams(UIUtils.dipToPx(36), UIUtils.dipToPx(36));
        rllpPb.setMarginStart(UIUtils.dipToPx(12));
        rllpPb.gravity = Gravity.CENTER_VERTICAL;
        pb.setLayoutParams(rllpPb);

        tv = new TextView(getContext());
        LinearLayout.LayoutParams rllpTv = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, UIUtils.dipToPx(36));
        rllpTv.setMarginStart(UIUtils.dipToPx(12));
        rllpTv.setMarginEnd(UIUtils.dipToPx(12));
        rllpTv.gravity = Gravity.CENTER_VERTICAL;
        tv.setLayoutParams(rllpTv);
        tv.setGravity(Gravity.CENTER_VERTICAL);
        tv.setTextSize(18f);
        addView(pb);
        addView(tv);

        TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.LoadingView, 0, 0);
        tv.setText(a.getString(R.styleable.LoadingView_text));
        tv.setTextColor(a.getColor(R.styleable.LoadingView_textColor, Color.BLACK));
        a.recycle();
    }
}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/comp/LoadingView.kt
package com.rarnu.tools.neo.comp

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.Gravity
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import com.rarnu.android.dip2px
import com.rarnu.tools.neo.R

class LoadingView : LinearLayout {

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initAttrs(attrs)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initAttrs(attrs)
    }

    constructor(context: Context) : super(context) {
        initAttrs(null)
    }

    private fun initAttrs(attrs: AttributeSet?) {
        val a = context.obtainStyledAttributes(attrs, R.styleable.LoadingView, 0, 0)
        orientation = HORIZONTAL
        background = context.resources.getDrawable(R.drawable.background_layout, context.theme)
        addView(ProgressBar(context).apply {
            layoutParams = LayoutParams(36.dip2px(), 36.dip2px()).apply {
                marginStart = 12.dip2px()
                gravity = Gravity.CENTER_VERTICAL
            }
        })
        addView(TextView(context).apply {
            layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, 36.dip2px()).apply {
                marginStart = 12.dip2px()
                marginEnd = 12.dip2px()
                gravity = Gravity.CENTER_VERTICAL
            }
            gravity = Gravity.CENTER_VERTICAL
            textSize = 18f
            text = a.getString(R.styleable.LoadingView_text)
            setTextColor(a.getColor(R.styleable.LoadingView_textColor, Color.BLACK))
        })
        a.recycle()
    }
}
